### PR TITLE
CB-14596 adding revert cases to upgrade ccm flow: reverting restore previous ccm connection

### PR DESCRIPTION
**Pull request by afarsang**
CB-14596 creating upgrade ccm error handling cases: ccm upgrade is creating a salt job which wait for a timeperiod, which redo the nginx config for the previous ccm config; this helps reestablish the connection with the machine even if the connection is lost during the ccm upgrade; the revert flow logic is waiting the same period of time in case of connection issues, but remove the salt job is the connection is in order

### CODE REVIEW

**Review by lajosrodek — CHANGES_REQUESTED** [review]
General note: The PR description is empty. Could you please copy (& adapt if necessary) the commit message over?

First slice of comments, considering changes up to `UpgradeCcmFailedEvent.java`. Some further may also come later.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ccm/UpgradeCcmOrchestratorService.java [hunk 1]
```diff
@@ -43,11 +45,17 @@ public class UpgradeCcmOrchestratorService {
     private HostOrchestrator hostOrchestrator;
 
     public void reconfigureNginx(Long stackId) throws CloudbreakOrchestratorException {
-        OrchestratorStateParams stateParams = createStateParams(stackId, NGINX_STATE);
+        OrchestratorStateParams stateParams = createStateParams(stackId, UPGRADECCM_STATE);
         LOGGER.debug("Calling reconfigureNginx with state params '{}'", stateParams);
         hostOrchestrator.runOrchestratorState(stateParams);
     }
 
+    public void finalize(Long stackId) throws CloudbreakOrchestratorException {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ccm/UpgradeCcmOrchestratorService.java:53]: Please cover with UT.

File: orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls [hunk 2]
```diff
@@ -0,0 +1,18 @@
+restore_last_ssl_conf:
+  module.run:
+    - name: file.restore_backup
+    - path: /etc/nginx/sites-enabled/ssl.conf
+    - backup_id: 0
+
+touch_ssl_conf_for_restart:
+  file.touch:
+    --name: /etc/nginx/sites-enabled/ssl.conf
```
Comment (lajosrodek) [inline orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls:9]: Sanity check: Is the `--name` possibly a typo? Was it rather meant as a list element `- name`?

File: orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls [hunk 3]
```diff
@@ -0,0 +1,18 @@
+restore_last_ssl_conf:
+  module.run:
+    - name: file.restore_backup
+    - path: /etc/nginx/sites-enabled/ssl.conf
+    - backup_id: 0
+
+touch_ssl_conf_for_restart:
```
Comment (lajosrodek) [inline orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls:7]: Perhaps add a dependency on `restore_last_ssl_conf`.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import com.sequenceiq.cloudbreak.reactor.api.event.cluster.upgrade.ccm.UpgradeCcmFinalizeRequest;
+import com.sequenceiq.cloudbreak.reactor.api.event.cluster.upgrade.ccm.UpgradeCcmFinalizeResult;
+import com.sequenceiq.flow.event.EventSelectorUtil;
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+import com.sequenceiq.flow.reactor.api.handler.HandlerEvent;
+
+import reactor.bus.Event;
+
+@Component
+public class FinalizeHandler extends ExceptionCatcherEventHandler<UpgradeCcmFinalizeRequest> {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java:22]: Please cover with UT.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandler.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.cloudbreak.core.flow2.cluster.ccm.upgrade.UpgradeCcmService;
+import com.sequenceiq.cloudbreak.reactor.api.event.cluster.upgrade.ccm.UpgradeCcmFailedEvent;
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+import com.sequenceiq.flow.reactor.api.handler.HandlerEvent;
+
+import reactor.bus.Event;
+
+@Component
+public class RevertAllHandler extends ExceptionCatcherEventHandler<UpgradeCcmFailedEvent> {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandler.java:24]: Please cover with UT.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.cloudbreak.core.flow2.cluster.ccm.upgrade.UpgradeCcmService;
+import com.sequenceiq.cloudbreak.reactor.api.event.cluster.upgrade.ccm.UpgradeCcmFailedEvent;
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+import com.sequenceiq.flow.reactor.api.handler.HandlerEvent;
+
+import reactor.bus.Event;
+
+@Component
+public class RevertSaltStatesHandler extends ExceptionCatcherEventHandler<UpgradeCcmFailedEvent> {
```
Comment (lajosrodek) [inline core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java:21]: Please cover with UT.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/cluster/upgrade/ccm/AbstractUpgradeCcmEvent.java [hunk 7]
```diff
@@ -1,27 +1,33 @@
...
 
-    public AbstractUpgradeCcmEvent(String selector, Long stackId, Long clusterId, Tunnel oldTunnel, Promise<AcceptResult> accepted) {
+    public AbstractUpgradeCcmEvent(String selector, Long stackId, Long clusterId, Tunnel oldTunnel, LocalDateTime revertTime, Promise<AcceptResult> accepted) {
         super(selector, stackId, accepted);
         this.clusterId = clusterId;
         this.oldTunnel = oldTunnel;
+        this.revertTime = revertTime;
     }
 
     public Long getClusterId() {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/cluster/upgrade/ccm/AbstractUpgradeCcmEvent.java:34]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/cluster/upgrade/ccm/AbstractUpgradeCcmEvent.java [hunk 8]
```diff
@@ -31,4 +37,8 @@ public Long getClusterId() {
     public Tunnel getOldTunnel() {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/cluster/upgrade/ccm/AbstractUpgradeCcmEvent.java:39]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/cluster/upgrade/ccm/UpgradeCcmFailedEvent.java [hunk 9]
```diff
@@ -1,34 +1,69 @@
...
+        this.clusterId = clusterId;
+        this.revertTime = revertTime;
+    }
+
+    @Override
+    public Long getClusterId() {
+        return clusterId;
     }
 
     public Tunnel getOldTunnel() {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/api/event/cluster/upgrade/ccm/UpgradeCcmFailedEvent.java:58]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    protected Selectable doAccept(HandlerEvent<UpgradeCcmFinalizeRequest> event) {
+        UpgradeCcmFinalizeRequest request = event.getData();
+        Long stackId = request.getResourceId();
+        LOGGER.info("Finalize CCM upgrade...");
+        try {
+            upgradeCcmService.finalize(stackId);
+        } catch (CloudbreakOrchestratorException e) {
+            LOGGER.error("Failed finalizing CCM upgrade with a salt state", e);
+        }
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java:51]: no, this is bug

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/ReconfigureNginxHandler.java [hunk 11]
```diff
@@ -23,9 +29,14 @@ public class ReconfigureNginxHandler extends ExceptionCatcherEventHandler<Upgrad
...
     private static final Logger LOGGER = LoggerFactory.getLogger(ReconfigureNginxHandler.class);
 
+    @Value("${cb.ccmRevertJob.activationInMinutes}")
+    private Integer activationInMinutes;
+
     @Inject
     private UpgradeCcmService upgradeCcmService;
 
+    private LocalDateTime revertDateTime;
+
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/ReconfigureNginxHandler.java:39]: right

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/ReconfigureNginxHandler.java [hunk 12]
```diff
@@ -34,21 +45,24 @@ public String selector() {
     @Override
     protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmReconfigureNginxRequest> event) {
         LOGGER.error("Reconfiguring NGINX for CCM upgrade has failed", e);
-        return new UpgradeCcmFailedEvent(resourceId, event.getData().getOldTunnel(), getClass(), e);
+        return new UpgradeCcmFailedEvent(UPGRADE_CCM_REVERT_SALTSTATE_EVENT.event(), resourceId, event.getData().getClusterId(), event.getData().getOldTunnel(),
+                getClass(), e, revertDateTime);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/ReconfigureNginxHandler.java:49]: no, it should have been the previous field (initiated, but was missing). i fixed it.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/ccm/upgrade/UpgradeCcmService.java [hunk 13]
```diff
@@ -115,13 +115,27 @@ void removeAgentState(Long stackId) {
         flowMessageService.fireEventAndLog(stackId, UPDATE_IN_PROGRESS.name(), ResourceEvent.CLUSTER_CCM_UPGRADE_REMOVE_AGENT);
     }
 
+    public void removeAgentFailed(Long stackId) {
+        String statusReason = "Remove previous version's agent from the cluster is failed";
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/ccm/upgrade/UpgradeCcmService.java:119]: "has been failed"

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/ccm/upgrade/UpgradeCcmService.java [hunk 14]
```diff
@@ -115,13 +115,27 @@ void removeAgentState(Long stackId) {
...
+
     void deregisterAgentState(Long stackId) {
         String statusReason = "Deregister previous version's agent";
         LOGGER.debug(statusReason);
         stackUpdater.updateStackStatus(stackId, DetailedStackStatus.CCM_UPGRADE_IN_PROGRESS, statusReason);
         flowMessageService.fireEventAndLog(stackId, UPDATE_IN_PROGRESS.name(), ResourceEvent.CLUSTER_CCM_UPGRADE_DEREGISTER_AGENT);
     }
 
+    public void deregisterAgentFailed(Long stackId) {
+        String statusReason = "Deregister previous version's agent is failed";
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/ccm/upgrade/UpgradeCcmService.java:133]: "has been failed"

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandler.java [hunk 15]
```diff
@@ -34,21 +34,23 @@ public String selector() {
...
         try {
             upgradeCcmService.removeAgent(stackId, request.getOldTunnel());
         } catch (CloudbreakOrchestratorException e) {
-            LOGGER.debug("Failed removing agent with a salt state");
-            return new UpgradeCcmFailedEvent(stackId, request.getOldTunnel(), getClass(), e);
+            LOGGER.error("Failed removing agent with a salt state", e);
+            upgradeCcmService.removeAgentFailed(stackId);
         }
-        return new UpgradeCcmRemoveAgentResult(stackId, request.getClusterId(), request.getOldTunnel());
+        return new UpgradeCcmRemoveAgentResult(stackId, request.getClusterId(), request.getOldTunnel(), request.getRevertTime());
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandler.java:54]: Yes, same as deregistering - the ccm upgrade is successful in this case. However I will add message about this.

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandler.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        Long clusterId = request.getClusterId();
+        LOGGER.info("Reverting all is starting");
+        upgradeCcmService.registerClusterProxy(stackId);
+        long ms = 0;
+        if (event.getData().getRevertTime() != null) {
+            ms = Duration.between(LocalDateTime.now(), event.getData().getRevertTime()).getSeconds() * MILLI;
+        } else {
+            LOGGER.warn("Revert time is not set!");
+        }
+        LOGGER.info("CCM upgrade milliseconds back to reverting {}ms", ms);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandler.java:57]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    private UpgradeCcmService upgradeCcmService;
+
+    @Override
+    public String selector() {
+        return UPGRADE_CCM_REVERT_SALTSTATE_COMMENCE_EVENT.event();
+    }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmFailedEvent> event) {
+        LOGGER.error("Reverting for CCM upgrade has failed", e);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java:35]: ok

File: orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls (hunk 2 above)
Comment (afarsang) [reply orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls:9]: yes, fixed

File: orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls (hunk 3 above)
Comment (afarsang) [reply orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls:7]: I couldn't figure out how can I do that

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java (hunk 4 above)
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java:22]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandler.java (hunk 5 above)
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandler.java:24]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java (hunk 6 above)
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java:21]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        return EventSelectorUtil.selector(UpgradeCcmFinalizeRequest.class);
+    }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmFinalizeRequest> event) {
+        LOGGER.error("Finalizing CCM upgrade has failed", e);
+        upgradeCcmService.removeAgentFailed(resourceId);
+        return new UpgradeCcmFailedEvent(resourceId, event.getData().getClusterId(), event.getData().getOldTunnel(), getClass(), e,
+                event.getData().getRevertTime());
+    }
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/FinalizeHandler.java:40]: I leave it there, but also put this in the try catch block to handle orchestartionExceptions

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandler.java [hunk 19]
```diff
@@ -34,21 +34,23 @@ public String selector() {
     @Override
     protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpgradeCcmRemoveAgentRequest> event) {
         LOGGER.error("Removing agent for CCM upgrade has failed", e);
-        return new UpgradeCcmFailedEvent(resourceId, event.getData().getOldTunnel(), getClass(), e);
+        upgradeCcmService.removeAgentFailed(resourceId);
+        return new UpgradeCcmRemoveAgentResult(resourceId, event.getData().getClusterId(), event.getData().getOldTunnel(),
+                event.getData().getRevertTime());
     }
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandler.java:40]: true, however this it what should happen when a very unlikely event happens :)

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/ReconfigureNginxHandlerTest.java [hunk 20]
```diff
@@ -51,19 +52,20 @@ void selector() {
...
+        ReflectionTestUtils.setField(underTest, "activationInMinutes", 5);
+        UpgradeCcmReconfigureNginxRequest request = new UpgradeCcmReconfigureNginxRequest(STACK_ID, CLUSTER_ID, Tunnel.CCM, null);
         when(event.getData()).thenReturn(request);
 
         Selectable result = underTest.doAccept(event);
         InOrder inOrder = inOrder(upgradeCcmService);
-        inOrder.verify(upgradeCcmService).updateTunnel(STACK_ID);
+        inOrder.verify(upgradeCcmService).updateTunnel(STACK_ID, Tunnel.latestUpgradeTarget());
         inOrder.verify(upgradeCcmService).reconfigureNginx(STACK_ID);
         assertThat(result.selector()).isEqualTo("UPGRADECCMRECONFIGURENGINXRESULT");
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/ReconfigureNginxHandlerTest.java:63]: good idea

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java [hunk 21]
```diff
@@ -51,31 +50,25 @@ void selector() {
...
         verify(upgradeCcmService).removeAgent(STACK_ID, Tunnel.CCM);
-        assertThat(result.selector()).isEqualTo("UPGRADECCMFAILEDEVENT");
-        assertThat(result).isInstanceOf(UpgradeCcmFailedEvent.class);
-        UpgradeCcmFailedEvent failedEvent = (UpgradeCcmFailedEvent) result;
-        assertThat(failedEvent.getOldTunnel()).isEqualTo(Tunnel.CCM);
-        assertThat(failedEvent.getResourceId()).isEqualTo(STACK_ID);
-        assertThat(failedEvent.getFailureOrigin()).isEqualTo(RemoveAgentHandler.class);
-        assertThat(failedEvent.getException().getMessage()).isEqualTo("salt error");
+        assertThat(result.selector()).isEqualTo("UPGRADECCMREMOVEAGENTRESULT");
     }
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:72]: ok

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java [hunk 22]
```diff
@@ -51,31 +51,30 @@ void selector() {
...
         when(event.getData()).thenReturn(request);
 
         Selectable result = underTest.doAccept(event);
         InOrder inOrder = inOrder(upgradeCcmService);
-        inOrder.verify(upgradeCcmService).updateTunnel(STACK_ID);
+        inOrder.verify(upgradeCcmService).updateTunnel(STACK_ID, Tunnel.latestUpgradeTarget());
         inOrder.verify(upgradeCcmService).removeAgent(STACK_ID, Tunnel.CCM);
+        assertThat(((UpgradeCcmRemoveAgentResult) result).getAgentDeletionSucceed().equals(Boolean.FALSE));
         assertThat(result.selector()).isEqualTo("UPGRADECCMREMOVEAGENTRESULT");
     }
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:63]: I wonder how this works; `getAgentDeletionSucceed()` was supposed to be set to `TRUE` here.

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java [hunk 23]
```diff
@@ -51,31 +51,30 @@ void selector() {
...
 
     @Test
     void orchestrationException() throws CloudbreakOrchestratorException {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:66]: Please also verify `upgradeCcmService.removeAgentFailed()` and `((UpgradeCcmRemoveAgentResult) result).getAgentDeletionSucceed() == TRUE`.

File: orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls (hunk 3 above)
Comment (lajosrodek) [reply orchestrator-salt/src/main/resources/salt/salt/nginx/restoressl.sls:7]: Not sure, perhaps add something like:
```
    - require:
        - module: restore_last_ssl_conf
```

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandlerTest.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+    void setUp() {
+    }
+
+    @Test
+    void selector() {
+        assertThat(underTest.selector()).isEqualTo("UPGRADE_CCM_REVERT_ALL_COMMENCE_EVENT");
+    }
+
+    @Test
+    void doAccept() {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandlerTest.java:53]: I wonder if it was somehow possible to verify the `Thread.sleep()` logic here.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Some further minor bits.

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java [hunk 25]
```diff
@@ -51,31 +51,30 @@ void selector() {
 
     @Test
     void doAccept() throws CloudbreakOrchestratorException {
```
Comment (lajosrodek) [inline core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:53]: Perhaps also verify that `upgradeCcmService.removeAgentFailed()` is never invoked in this case.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java (hunk 25 above)
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:53]: removeAgentFailed is called in this case too

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java (hunk 23 above)
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:66]: ok

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java (hunk 22 above)
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:63]: added test for it

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java [hunk 26]
```diff
@@ -51,31 +51,30 @@ void selector() {
...
         doThrow(new CloudbreakOrchestratorFailedException("salt error")).when(upgradeCcmService).removeAgent(any(), any());
 
         Selectable result = underTest.doAccept(event);
         verify(upgradeCcmService).removeAgent(STACK_ID, Tunnel.CCM);
-        assertThat(result.selector()).isEqualTo("UPGRADECCMFAILEDEVENT");
-        assertThat(result).isInstanceOf(UpgradeCcmFailedEvent.class);
-        UpgradeCcmFailedEvent failedEvent = (UpgradeCcmFailedEvent) result;
+        assertThat(result.selector()).isEqualTo("UPGRADECCMREMOVEAGENTRESULT");
+        assertThat(result).isInstanceOf(UpgradeCcmRemoveAgentResult.class);
+        UpgradeCcmRemoveAgentResult failedEvent = (UpgradeCcmRemoveAgentResult) result;
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:75]: kinda yes, but I'll figure outbetter name for this

File: core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertSaltStatesHandler.java:35]: sorry, I though I did

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandlerTest.java (hunk 24 above)
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandlerTest.java:53]: I though this could be overkill

Comment (lajosrodek) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RevertAllHandlerTest.java:53]: Sounds reasonable. Can be still added later if needed.

**Review by lajosrodek — CHANGES_REQUESTED** [review]
Thanks. A few more outstanding issues:

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java [hunk 27]
```diff
@@ -51,31 +51,33 @@ void selector() {
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:51]: Still pending, but also okay to do this later.

File: core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java (hunk 25 above)
Comment (lajosrodek) [reply core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/upgrade/ccm/RemoveAgentHandlerTest.java:53]: Still unresolved.

**Review by lajosrodek — APPROVED** [review]
Only two minor bits, but LGTM otherwise.

**Pull request merged**